### ISSUE TITLE
possible memory leak with #connection.useDb

### ISSUE THREAD

**Issue by drish** [Automattic/mongoose#2073]
Hello guys,

In my work, I need to switch through various different databases.
I faced a memory leak warning when using `.useDb` several times with the same dabatabase name. 

The following code demostrates this issue.

``` js
var mongoose = require('mongoose');

var db1 = mongoose.createConnection('mongodb://localhost/my_database');

for (var i = 0; i <= 11; i++){
  var db2 = db1.useDb('test')
}
```

``` sh
$ node -v 
v0.10.28
```

mongoose version - latest (3.8.9)
### Possible fix

I could fix it locally, by looping through `db1.otherDbs` object, and returning the connection if it already exists, otherwise it triggers the memory leak.
Maybe , `useDb` function should return an existing db instance, if it already exists in the caller `caller_db.otherDbs` ? If so, I can pull request that.

**Comment by bmpc** [comment]
Hi guys,

I think a lot has changed since 2014. From what I understand, connections are now being cached by default when using ```useDb()```.

However, I believe that in our case we are having memory leaks related to this. 

We have a multi-tenant architecture where each tenant has its own database. Because we have many tenants (+ 3000) I think these ```NativeConnection``` objects might be originating this memory leak. 

We have tried to disable the ```relatedDbs``` cache with 

`mongoose.connection.useDb("tenantDB*", { useCache: false });`

and manage the cache ourselves outside Mongoose. 

However, it seems that all connections are still being cached by the ```otherDbs``` array on the original connection. 

Is it possible to completely remove the ```otherDds``` array as this is being replaced with ```relatedDbs```?

Thank you.
Bruno.

**Comment by bmpc** [comment]
Hi guys,

I think a lot has changed since 2014. From what I understand, connections are now being cached by default when using ```useDb()```.

However, I believe that in our case we are having memory leaks related to this. 

We have a multi-tenant architecture where each tenant has its own database. Because we have many tenants (+ 3000) I think these ```NativeConnection``` objects might be originating this memory leak. 

We have tried to disable the ```relatedDbs``` cache with 

`mongoose.connection.useDb("tenantDB*", { useCache: false });`

and manage the cache ourselves outside Mongoose. 

However, it seems that all connections are still being cached by the ```otherDbs``` array on the original connection. 

Is it possible to completely remove the ```otherDds``` array as this is being replaced with ```relatedDbs```?

Thank you.
Bruno.

**Comment by vkarpov15** [comment]
@bmpc I opened a separate issue to track :point_up:

**Comment by vkarpov15** [comment]
@bmpc I took a closer look and there's no way to remove `otherDbs` without fundamentally breaking `useDb()`. That's because each connection created by `useDb()` needs to be tied to the original, because state changes in the original connection should propagate to the `useDb()` connection.

I have 2 recommendations:

1) Use `useCache: true`. That prevents Mongoose from creating a new connection every time you call `useDb()`.
2) If that causes problems and you want to cache yourself, use [`createConnection()`](https://mongoosejs.com/docs/connections.html#multiple_connections) instead of `useDb()` to create a new standalone connection and cache these connections yourself. We generally recommend this approach for multi-tenant to avoid [one tenant's slow trains affecting another tenant](https://thecodebarbarian.com/slow-trains-in-mongodb-and-nodejs).

**Comment by bmpc** [comment]
Hi @vkarpov15,

Meanwhile, we have identified the "real" cause of our leaks. It was related to the mongoose connections but they were not the root cause of the issue.

It all comes down to Domains in nodejs and the following:

> If domains are in use, then all new EventEmitter objects (including Stream objects, requests, responses, etc.) will be implicitly bound to the active domain at the time of their creation.
from https://nodejs.org/api/domain.html#domain_implicit_binding

We were using "Raven" to send logs to Sentry. As it happens, this module was creating a domain for each request in which all the request information was being tied to the domain object.

Because Mongoose connections inherit from EventEmitter, they were also using the same Domain object. Therefore, because all Mongoose connections are cached, the domain and consequent request information was never collected.

We have removed the "Raven" module for now, and memory is now controlled.

Either way, thank you for considering this. I had also noticed that there was some "connection state" that was being propagated across all Mongoose Connections and I was not sure if we could simply remove BDs from memory.

Best Regards,
Bruno.
